keps/sig-storage/596-csi-inline-volumes/README.md

```diff
@@ -213,32 +250,214 @@ Inline volume requests can only participate in mount/unmount volume operations.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+  Check if any CSIDriver specs have `VolumeLifecycleModes` including an
+  "Ephemeral" `VolumeLifecycleMode` and if any pod specs are using this driver inline.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Other (treat as last resort)
+  - Details:
+  For pods using an inline CSI driver (as in [Examples](#examples)), review
+  the status of the pod (did it mount and start successfully?).
+  Additionally, the cluster admin may review the kubelet logs for mount/unmount calls.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+- No increased failure rates during mount operations.
+- Mount times should be expected to be less than or equal to the default behavior.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Metrics
+  - Metric name: storage_operation_duration_seconds
+    - [Optional] Aggregation method: filter by `operation_name = volume_mount`
+      - Components exposing the metric: kubelet
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+  No additional metrics needed.
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+  There are no new dependencies introduced by this feature.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+  There will be no new API calls. A new API field is added to the CSIDriver
+  API to determine if the driver supports ephemeral inline volumes.
+
+###### Will enabling / using this feature result in introducing new API types?
+
+  The only new API type is the `VolumeLifecycleMode` string that is used
+  to specify supported modes in the CSIDriver spec.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+  No, ephemeral inline volumes can only participate in mount/unmount volume operation
+  calls to the CSI driver.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+  There is a new `VolumeLifecycleModes` field in CSIDriverSpec that may slightly increase
+  the size by adding 2 new strings (`Persistent` and `Ephemeral`).
+
+  There is also `pod.spec.volumes` that can contain `csi` items now. Its size is comparable
+  to the size of Secrets/ConfigMap volumes inline in Pods. The CSI volume definitions will
+  be bigger, as they contain a generic map of `volumeAttributes`, which contains opaque
+  parameters for the CSI driver.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+  Compared to CSI persistent volumes, there should be no increase in the amount of time
```

Comment:
nit: it's worth clarifying that mount time is on a critical part for pod-startup slo:
https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md

And configmaps/secrets are actually (at least as of now) part of stateless pods, which means that we may see a regression.
So following up on that (wearing my scalability hat now) - do we have any numbers about this now?

Reply:
> nit: it's worth clarifying that mount time is on a critical part for pod-startup slo:

I've added a note to this in the KEP with a link to the [pod startup latency SLO](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md).

> And configmaps/secrets are actually (at least as of now) part of stateless pods, which means that we may see a regression.

Ah, but the definition in the [pod startup latency SLO](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md) states:
```
[2]A stateless pod is defined as a pod that doesn't mount volumes with sources other than secrets, config maps, downward API and empty dir.
```
If a pod uses CSI inline volumes, that is still a type of volume. So it's not really a stateless pod in that case.

> So following up on that (wearing my scalability hat now) - do we have any numbers about this now?

No, I don't have any numbers readily available. What kind of comparison would be interesting to see?

Reply:
> If a pod uses CSI inline volumes, that is still a type of volume. So it's not really a stateless pod in that case.

OK - you're right. Thanks for correcting me.

> No, I don't have any numbers readily available. What kind of comparison would be interesting to see?

So what I'm looking for is a comparison, that e.g. if I have a pod with 5 secrets, it takes X seconds now, but it takes Y second if we mount those secrets via csi inline volume.

[Again - non-blocking, as it's not a replacement, it's a new way of doing stuff. I'm asking more for building awareness about that.]

Reply:
> So what I'm looking for is a comparison, that e.g. if I have a pod with 5 secrets, it takes X seconds now, but it takes Y second if we mount those secrets via csi inline volume.

Ok, I'd expect this to vary based on the CSI driver used, but I agree it would be useful to document this comparison so we at least have some idea of how big that difference is for a basic use case like this to set expectations for users of this feature. I've added a TODO line in this section to follow up on it during GA testing.

Reply:
Thanks!